controller/kongplugininstallation/image/image.go

```diff
@@ -0,0 +1,142 @@
+					mut.Lock()
+					layersThatMayContainPlugin = append(layersThatMayContainPlugin, desc)
+					mut.Unlock()
+				}
+				return nil
+			},
+		},
+	}); err != nil {
+		return nil, fmt.Errorf("can't fetch image: %s, because: %w", imageURL, err)
+	}
+	// How to build such image is described in details,
+	// following manual results in the image with exactly one layer that contains a plugin.
+	if numOfLayers := len(layersThatMayContainPlugin); numOfLayers != 1 {
+		return nil, fmt.Errorf("expected exactly one layer with plugin, found %d layers", numOfLayers)
+	}
+	layerWithPlugin := layersThatMayContainPlugin[0]
+	contentOfLayerWithPlugin, err := inMemoryStore.Fetch(ctx, layerWithPlugin)
+	if err != nil {
+		return nil, fmt.Errorf("can't get layer of image: %w", err)
+	}
+
+	return extractKongPluginFromLayer(contentOfLayerWithPlugin)
+}
+
+// CredentialsStoreFromString expects content of typical configuration as a string, described
+// in https://kubernetes.io/docs/tasks/configure-pod-container/pull-image-private-registry
+// and returns credentials.Store.
+// This is typical way how private registries are used with Docker and Kubernetes.
+func CredentialsStoreFromString(s string) (credentials.Store, error) {
+	// TODO: Now we create temporary file, which is not great and should be changed,
+	// but it's the only way to use credentials.NewFileStore(...) which robustly
+	// parses config.json (format used by Docker and Kubernetes).
+	tmpFile, err := os.CreateTemp("", "credentials")
+	if err != nil {
+		return nil, fmt.Errorf("failed to create temporary file: %w", err)
+	}
+	defer tmpFile.Close()
+	defer os.Remove(tmpFile.Name())
+	if _, err = tmpFile.WriteString(s); err != nil {
+		return nil, fmt.Errorf("failed to write credentials to file: %w", err)
+	}
+	return credentials.NewFileStore(tmpFile.Name())
+}
+
+func extractKongPluginFromLayer(r io.Reader) ([]byte, error) {
+	gr, err := gzip.NewReader(r)
+	if err != nil {
+		return nil, fmt.Errorf("failed to parse layer as tar.gz: %w", err)
+	}
+
+	// The target file name for custom Kong plugin.
+	const kongPluginName = "plugin.lua"
+
+	// Search for the file walking through the archive.
+	// Limit plugin to 1MB the same as ConfigMap in Kubernetes.
+	const sizeLimit_1MiB = 1024 * 1024
+
+	for tr := tar.NewReader(io.LimitReader(gr, sizeLimit_1MiB)); ; {
+		switch h, err := tr.Next(); {
+		case err == nil:
+			if filepath.Base(h.Name) == kongPluginName {
+				plugin := make([]byte, h.Size)
+				if _, err := io.ReadFull(tr, plugin); err != nil {
+					return nil, fmt.Errorf("failed to read %s from image: %w", kongPluginName, err)
+				}
+				return plugin, nil
+			}
+		case errors.Is(err, io.EOF):
+			return nil, fmt.Errorf("file %q not found in the image", kongPluginName)
+		case errors.Is(err, io.ErrUnexpectedEOF):
+			return nil, fmt.Errorf("plugin size exceed %d bytes", sizeLimit_1MiB)
+		default:
+			return nil, fmt.Errorf("unexpected error during looking for plugin: %w", err)
+		}
+	}
+}
```

Comment:
Can we add a unit test for this function to cover the edge/error cases (no `plugin.lua` in the archive, `plugin.lua` bigger than 1MB, etc.)?